File: FastGFX/js/SpritesTexture.js
var STNode=function(x,y,width,height) {
  this.left = null
  this.right = null
  this.image = null
  this.x = x
  this.y = y
  this.width = width
  this.height = height
}
STNode.prototype.insert = function(image,spacing) {
  var box_width=image.width + spacing
  var box_height=image.height + spacing

  if(this.left && this.right) {
    return this.left.insert(image,spacing) || this.right.insert(image,spacing)
  } else {
    // if there's already a sprite here, return
    if(this.image) return null;
    // if we're too small, return
    if((box_width>this.width) || ((box_height)>this.height)) return null
    if((box_width==this.width) && ((box_height)==this.height)) { //if we're just right, accept
      this.image=image
      return this
    } else { // otherwise, gotta split this node and create some kids
      var dw = this.width - (box_width)
      var dh = this.height - (box_height)
      if(dw>dh) {
        this.left = new STNode(this.x, this.y, box_width, this.height)
        this.right = new STNode(this.x+box_width, this.y, this.width-(box_width), this.height)
      } else {
        this.left = new STNode(this.x, this.y, this.width, box_height)
        this.right = new STNode(this.x, this.y+box_height, this.width, this.height-(box_height))
      }
      // insert into first child we created
      return this.left.insert(image,spacing)
    }
  }
}

var SpritesTexture = function(gl,size,spacing) {
  this.gl = gl
  this.size = size

  this.canvas = document.createElement('canvas')
  this.canvas.width = this.size
  this.canvas.height = this.size
  this.ctx = this.canvas.getContext('2d')

  this.sprites = []
  this.spritesByName = {}

  this.spacing = spacing
  this.needUpload = false

  var gl = this.gl
  this.texture = gl.createTexture()
  gl.bindTexture(gl.TEXTURE_2D, this.texture)
  gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR)
  gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR_MIPMAP_LINEAR)
  gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
  gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);

  while(GL.textures.length == 0) GL.textures.push(null)
  this.textureId = GL.textures.length
  GL.textures.push(this.texture)

}
SpritesTexture.prototype.tryPack = function(sprites) {
  var nodes = new Array(sprites.length)
  var root = new STNode(0,0,this.size,this.size)

  sprites = sprites.sort((function(a,b) {
    return Math.max(b.width,b.height)-Math.max(a.width,a.height)
  }).bind(this))

  for(var i=0; i< sprites.length; i++) {
    var sprite = sprites[i]
    var node = root.insert(sprite.image,this.spacing*2)
    if(!node) return false;
    nodes[i] = node
  }

  var fix = -2

  for(var i=0; i<sprites.length; i++) {
    var node = nodes[i]
    var sprite = sprites[i]
    sprite.coords = {
      width: node.width,
      height: node.height,
      x: node.x,
      y: node.y,
      xmin: (node.x+fix)/this.size,
      ymin: (node.y+fix)/this.size,
      xmax: (node.x+node.width-this.spacing+fix)/this.size,
      ymax: (node.y+node.height-this.spacing+fix)/this.size
    }
  }
  this.sprites = sprites
  this.needUpload = true
  return true
}
SpritesTexture.prototype.tryAddSprite = function(sprite) {
  if(!this.tryPack(this.sprites.concat([sprite]))) {
    return false;
  } else {
    return true;
  }
}
SpritesTexture.prototype.isBigEnough = function(w,h) {
  return (w<this.size && h<this.size)
}
SpritesTexture.prototype.upload = function() {
  if(!this.needUpload) return;

  var ctx = this.ctx
  var canvas = this.canvas
  var gl = this.gl
  ctx.clearRect(0, 0, canvas.width, canvas.height)

  /*ctx.fillStyle="#FF00FF"
  ctx.fillRect(0,0,canvas.width, canvas.height)*/

  for(var i=0; i<this.sprites.length; i++) {
    var sprite = this.sprites[i]
    ctx.drawImage(sprite.image,sprite.coords.x,sprite.coords.y)
  }

  gl.bindTexture(gl.TEXTURE_2D, this.texture)
  gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, gl.RGBA, gl.UNSIGNED_BYTE, canvas)
  gl.generateMipmap(gl.TEXTURE_2D)

  for(var i=0; i<this.sprites.length; i++) {
    var sprite = this.sprites[i]
    sprite.texture = this.textureId
    //console.log("COORDS",sprite.name,sprite.coords)
    Module._fgfx_Sprite_setTextureFragment(sprite.sprite, sprite.texture,
      sprite.coords.xmin, sprite.coords.ymin, sprite.coords.xmax, sprite.coords.ymax,
      sprite.coords.width, sprite.coords.height, false)
  }

  this.needUpload = false
}
SpritesTexture.prototype.clean = function(engineTime) {
  for(var i=0; i<this.sprites.length; i++) {
    var sprite = this.sprites[i].sprite
    if(sprite && engineTime-sprite.lastUseTime>30) { /// REMOVE SPRITE
      console.log("REMOVE SPRITE",this.sprites[i].name)
      sprite.unload()
      this.sprites.splice(i,1)
      i--;
    }
  }
}

export { SpritesTexture }
export default SpritesTexture

File: FastGFX/src/SpritesTexture.cpp
//
// Created by Michał Łaszczewski on 09/12/16.
//

#ifndef EMSCRIPTEN

#include <vector>
#include <algorithm>
#include <fastgfx.h>

namespace fgfx {

  STNode::STNode(int xp, int yp, int widthp, int heightp) : left(nullptr), right(nullptr), image(nullptr),
    x(xp), y(yp), width(widthp), height(heightp) {
  }
  std::shared_ptr<STNode> STNode::insert(std::shared_ptr<Image> src_image, int spacing) {
    int box_width = src_image->width + spacing;
    int box_height = src_image->height + spacing;

    if (left && right) {
      std::shared_ptr<STNode> r = left->insert(src_image, spacing);
      if(r) return r;
      return right->insert(src_image, spacing);
    } else {
      // if there's already a sprite here, return
      if (image) return nullptr;
      // if we're too small, return
      if ((box_width > width) || ((box_height) > height)) return nullptr;
      if ((box_width == width) && ((box_height) == height)) { //if we're just right, accept
        image = src_image;
        return shared_from_this();
      } else { // otherwise, gotta split this node and create some kids
        int dw = width - box_width;
        int dh = height - box_height;
        if (dw > dh) {
          left = std::make_shared<STNode>(x, y, box_width, height);
          right = std::make_shared<STNode>(x + box_width, y, width - box_width, height);
        } else {
          left = std::make_shared<STNode>(x, y, width, box_height);
          right = std::make_shared<STNode>(x, y + box_height, width, height - box_height);
        }
        // insert into first child we created
        return left->insert(src_image, spacing);
      }
    }
  }

  SpritesTexture::SpritesTexture(int sizep, int spacingp) : size(sizep), spacing(spacingp) {
    image = std::make_shared<Image>(size, size);
    textureCreated = false;
  }

  bool SpritesTexture::tryPack(std::vector<SpriteInfo> &sprites) {
    std::vector<std::shared_ptr<STNode>> nodes(sprites.size());

    std::shared_ptr<STNode> root = std::make_shared<STNode>(0, 0, size, size);

    std::sort(sprites.begin(), sprites.end(),
              [](const SpriteInfo & a, const SpriteInfo & b) -> bool {
                return std::max(b.image->width, b.image->height) < std::max(a.image->width, a.image->height);
              });

    for (int i = 0; i < sprites.size(); i++) {
      SpriteInfo sprite = sprites[i];
      std::shared_ptr<STNode> node = root->insert(sprite.image, spacing*2);
      if (!node) return false;
      nodes[i] = node;
    }


    /// TODO: posprzątac
    float fix = -2;

    for (int i = 0; i < sprites.size(); i++) {
      std::shared_ptr<STNode>& node = nodes[i];
      SpriteInfo& sprite = sprites[i];
      sprite.x = node->x;
      sprite.y = node->y;
      sprite.xmin = (1.0 * node->x + fix) / size;
      sprite.xmax = (1.0 * node->x + node->width + fix - spacing) / size;
      sprite.ymin = (1.0 * node->y + fix) / size;
      sprite.ymax = (1.0 * node->y + node->height + fix - spacing) / size;
      sprite.texture = texture;
    }
    packedSprites = sprites;

    image->clear();
    for(SpriteInfo& info : sprites) {
      image->put(info.image, info.x, info.y);
    }

    fgfx_log("SPRITES TEXTURE REQUEST UPLOAD")
    fgfx::engine->uploadQueue.enqueue([this]{
      this->upload();
    });

    return true;
  }

  bool SpritesTexture::tryAddSprite(SpriteInfo& sprite) {
    std::vector<SpriteInfo> newSprites = packedSprites;
    newSprites.push_back(sprite);
    return tryPack(newSprites);
  }

  bool SpritesTexture::isBigEnough(int w, int h) {
    return (w < size && h < size);
  }

  void SpritesTexture::clean() {
    for (int i = 0; i < packedSprites.size(); i++) {
      SpriteInfo& info = packedSprites[i];
      if (info.sprite && (!info.sprite->locked) && (!info.sprite->locked)
          && fgfx::engine->renderTime - info.sprite->lastUseTime > 3000) { /// REMOVE SPRITE
        info.sprite->unload();
        packedSprites.erase(packedSprites.begin()+i);
        i--;
      }
    }
  }

  void SpritesTexture::upload() { // in renderer thread
    fgfx_log("SPRITES TEXTURE UPLOAD")
    if(!textureCreated) {
      glGenTextures(1, &texture);
      glBindTexture(GL_TEXTURE_2D, texture);
      glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
      glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);
      glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
      glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
      textureCreated = true;
    }

    glBindTexture(GL_TEXTURE_2D, texture);
    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, image->width, image->height, 0, GL_RGBA, GL_UNSIGNED_BYTE, image->data);
    glGenerateMipmap(GL_TEXTURE_2D);

    for(SpriteInfo& info : packedSprites) {
      info.sprite->setTextureFragment(texture, info.xmin, info.ymin, info.xmax, info.ymax,
                                      info.image->width, info.image->height, false);
    }
  }

}


#endif
